File: validation/poc-00/java-candidate/src/main/java/dev/vra/poc00/domain/OrderState.java
package dev.vra.poc00.domain;

public enum OrderState {
    CREATED, PENDING_PAYMENT, CONFIRMED, CANCELLED, EXPIRED;

    public void requireTransitionTo(OrderState next) {
        boolean allowed = switch (this) {
            case CREATED -> next == PENDING_PAYMENT;
            case PENDING_PAYMENT -> next == CONFIRMED || next == CANCELLED || next == EXPIRED;
            case CONFIRMED, CANCELLED, EXPIRED -> false;
        };
        if (!allowed) throw new DomainFailure.InvalidTransition();
    }
}

File: validation/poc-00/kotlin-candidate/src/main/kotlin/dev/vra/poc00/kotlin/domain/OrderState.kt
package dev.vra.poc00.kotlin.domain

enum class OrderState {
    CREATED, PENDING_PAYMENT, CONFIRMED, CANCELLED, EXPIRED;

    fun requireTransitionTo(next: OrderState) {
        val allowed = when (this) {
            CREATED -> next == PENDING_PAYMENT
            PENDING_PAYMENT -> next == CONFIRMED || next == CANCELLED || next == EXPIRED
            CONFIRMED, CANCELLED, EXPIRED -> false
        }
        if (!allowed) throw DomainFailure.InvalidTransition()
    }
}
